kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/KopRecordsUtil.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.kop.utils;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.common.record.AbstractRecords;
+import org.apache.kafka.common.record.ConvertedRecords;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.record.TimestampType;
+
+/**
+ * Utils for DownConverted and ReAssignOffset operations.
+ */
+@Slf4j
+public class KopRecordsUtil {
+
+    public static ConvertedRecords<MemoryRecords> convertAndAssignOffsets(Iterable<? extends RecordBatch> batches,
+                                                              byte toMagic,
+                                                              long firstOffset) {
+        // maintain the batch along with the decompressed records to avoid the need to decompress again
+        List<RecordBatchAndRecords> recordBatchAndRecordsList = new ArrayList<>();
+        int totalSizeEstimate = 0;
+
+        long batchStartOffset = firstOffset;
+        for (RecordBatch batch : batches) {
+            if (toMagic < RecordBatch.MAGIC_VALUE_V2 && batch.isControlBatch()) {
+                continue;
+            }
```

Comment:
Could you migrate the logic from latest Kafka source code? For example, batch with ZSTD compression is not supported.

```java
                if (batch.compressionType() == CompressionType.ZSTD)
                    throw new UnsupportedCompressionTypeException("Down-conversion of zstandard-compressed batches " +
                        "is not supported");
```

Reply:
And handle the `batch.magic() <= toMagic` case

Reply:
> Could you migrate the logic from latest Kafka source code? For example, batch with ZSTD compression is not supported.
> 
> ```java
>                 if (batch.compressionType() == CompressionType.ZSTD)
>                     throw new UnsupportedCompressionTypeException("Down-conversion of zstandard-compressed batches " +
>                         "is not supported");
> ```

Thanks for your review. I have addressed your comment. PTAL

Reply:
> And handle the `batch.magic() <= toMagic` case

Since the internal message set of magic 0 and magic 1 has internal offset, the internal offset may be destroyed, so we still need to deal with the message of batch.magic <= toMagic. The only thing that remains unchanged is to ensure that the magic remains unchanged.